Project-9 Merge to Main

Brian-J-Wang opened the pull request:
Added API calls to the server and further polished the website

Brian-J-Wang committed:
feat: connecting webpage to server using the new API class
    - likes are now saved
    - cards are now saved
    - new cards are saved to the server
feat: prior to deleting a new popupWithConfirm class is used
    - deleting cards are now removed from the server

Brian-J-Wang committed:
feat: modal forms now show that data is being sent to the server when user submits the form

Brian-J-Wang committed:
package setup

gennady-bars reviewed (commented):
Dear Brian, (**Please, expand the general comment ↓**)

Unfortunately I have to reject the project without reviewing:



* There are errors in the console http://joxi.ru/BA0YjEvSqeO3Jr 
* When I like/dislike **new** cards there are errors http://joxi.ru/Rmzy1GkCVoB4jr 
* When I delete **new** cards there are errors http://joxi.ru/VrwxXvlTgn9E5m 
* When I change the avatar there is an error
* When I reload the page all newly deleted cards appear again at the bottom of the page


I've noticed another mistake:



* You need to **`disable`** the submit button in the `card popup` if the inputs are **empty** when you open it. It’s needed because I can click the button and add an empty card but it should **not** happen. After the first submit the button gets activated and you need to **deactivate** it again. Even if I can't add an empty card the button should look **disabled** (gray) anyway and with attribute `disabled`.
*

Brian-J-Wang committed:
fix: API class errors when working on a new workspace

Brian-J-Wang committed:
fix: removed causes of console errors from html file

Brian-J-Wang committed:
fix: the edit avatar modal shows disabled when there are no inputs

gennady-bars reviewed (commented):
Dear student, (**Please, expand the general comment ↓**)

Unfortunately I have to reject the project without reviewing:


* When I like/dislike **new** cards there are errors http://joxi.ru/Rmzy1GkCVoB4jr 
* When I delete **new** cards there are errors http://joxi.ru/VrwxXvlTgn9E5m 
* When I reload the page all newly deleted cards appear again at the bottom of the page

Brian-J-Wang commented:
good morning,

Can you please explain the steps you took to get an error for:
- liking/ disliking cards 
- deleting cards? 
I cannot replicate it on my end.

Brian-J-Wang committed:
fix: card id not saved when creating a new card

gennady-bars commented:
> good morning,
> 
> Can you please explain the steps you took to get an error for:
> 
> * liking/ disliking cards
> * deleting cards?
>   I cannot replicate it on my end.

Hi, Brian

1. Add 3 **new** cards
2. Try to like/dislike them (the **new** cards)
3. Try to delete them
4. And then reload the page to see all deleted cards appear again


But now everything is fine. You've corrected it

gennady-bars commented on src/Components/Api.js:
```diff
@@ -18,11 +18,14 @@ class Api {
+            })
         })
         .then (res => {
             if (res.ok) {
```
```js
.then(res => {
                if (res.ok) {
                    return res.json();
                }
                return Promise.reject(`Error ${res.status}`);
            })

```

The code of `ok` checking is repeated in **every** request of the project. To avoid it you need to make a special method `_checkResponse` that will check any response from the server if it’s ok or not.


```js
_checkResponse(res) {
        // here is the code of the checking
    }

```

And now you can replace redundant repeating with the only line:


```js
 .then(this._checkResponse)

```

Please, pay attention that you need to pass only the **reference** for the method rather than the call. 

**Could be improved**

You can make a special method (function ) for fetching and checking responses **not to duplicate** it in every request:

```js

  _request(url, options) {
    return fetch(url, options).then(this._checkResponse)
  }

```

And now you can safely replace `fetch` with `this._request` and remove repeated checking. And you don’t need to change anything else.

gennady-bars commented on src/Components/Api.js:
```diff
@@ -55,8 +85,46 @@ class Api {
+        })
     }
 
     handleError(res) {
```
I don't see some code of class `Api` . Actually you shouldn't have merged the changed of the 9th sprint into `main` unless the project has been accepted



```js

        .catch(err => {
            console.error(err);
        })
```


It’s better not to use `catch` in `Api` because you need to `catch` possible errors at the very **end** of any request and this end is in `index.js`   after all blocks `then`. And in real projects blocks `catch` usually open error popups and so on.

gennady-bars commented on src/Components/Card.js:
```diff
@@ -30,13 +36,21 @@ class Card {
 
     _setLikeHandler(cardCopy) {
         this._heartButton.addEventListener("click", () => {
             this._heartButton.classList.toggle('places__heart_enabled');
```
You need to **toggle likes** only in block `then` after a successful response from the server because if there is an error the user will not know about that and will see the toggled heart color.

gennady-bars commented on src/Components/FormValidator.js:
```diff
@@ -19,11 +19,14 @@ class FormValidator {
         });
 
+        this._formButton.addEventListener("click", () => {
+             this._toggleButtonState();
```
Please, delete it

You need to `disable` the button only after **resetting** the form when you have submitted a card in block `then`  to be able to click it again if there is a server error

gennady-bars commented on src/Components/PopupWithConfirm.js:
```diff
@@ -0,0 +1,36 @@
+        this._popupForm.addEventListener('submit', (evt) => {
+            evt.preventDefault();
+
+            this._popupSubmitButton.textContent = 'Saving...';
```
`'Saving...'`

When you delete a card you need `Deleting...`

gennady-bars commented on src/Components/PopupWithConfirm.js:
```diff
@@ -0,0 +1,36 @@
+                console.error(reject);
+            })
+            .finally(() => {
+                this.close();
```
` this.close();`

You need to close popups only in blocks `then` after a **successful response** from the server so that when there is an error it would be opened (the inputs will not be cleared and the user could send the data again to the server) 

Add `then` in line 19

gennady-bars commented on src/Components/PopupWithConfirm.js:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    onPopupConfirm(onConfirm) {
+        this.onConfirm = onConfirm;
```
nice function updating

gennady-bars commented on src/Components/PopupWithForm.js:
```diff
@@ -22,9 +24,17 @@ export default class PopupWithForm extends Popup {
+                console.error(reject);
+            })
+            .finally(() => {
+                this.close();
```
` this.close();`

You need to close popups only in blocks `then` after a **successful response** from the server so that when there is an error it would be opened (the inputs will not be cleared and the user could send the data again to the server) 

Add `then` in line 30

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,10 +1,12 @@
 import PopupWithImage from "../Components/PopupWithImage.js";
+import PopupWithConfirm from "../Components/PopupWithConfirm.js";
 
 const config = {
```
From the 8th sprint all constants (`initialCards` and `config`) should be in file `constants.js` in folder `utils` to clean up the main file.

gennady-bars commented on src/pages/index.js:
```diff
@@ -15,42 +17,44 @@ const config = {
+        const cardElement = renderCard(card);
+        container.addItem(cardElement);
+    });
+});
```
You need to `catch` possible errors at the very **end** of any server request.

**Please, correct it everywhere**

gennady-bars commented on src/pages/index.js:
```diff
@@ -67,55 +71,84 @@ imagePopup.setEventListeners();
+const profileElement = new UserInfo({name: nameSelector, desc: descSelector, avatar: avatarSelector});
+API.getUserInformation()
+    .then(userInfo => {
+        document.querySelector(avatarSelector).setAttribute('src', userInfo.avatar);
```
Class `UserInfo` should set the entire user’s data (including the **avatar**)

gennady-bars commented on src/pages/index.js:
```diff
@@ -67,55 +71,84 @@ imagePopup.setEventListeners();
+}
+
+const avatarModalForm = document.forms['editAvatarModalForm']
+const avatarModalValidator = new FormValidator(avatarModalForm, config, editAvatarSelector);
```
**Could be improved**

You can **universally** create instances of validators for all forms in the project storing them inside one object `formValidators`.  And then you can take any validator using attribute `name` of the form where you need to disable/enable the submit button or remove errors.

```js
// define an object for storing validators
const formValidators = {}

const enableValidation = (config) => {
  const formList = Array.from(document.querySelectorAll(config.formSelector))
  formList.forEach((formElement) => {
    const validator = new FormValidator(config, formElement)
    // here you get the name of the form (if you don’t have it then you need to add it into each form in `index.html` first)
    const formName = formElement.getAttribute('name')

   // here you store the validator using the `name` of the form
    formValidators[formName] = validator;
    validator.enableValidation();
  });
};

enableValidation(config);

```

And now you can use them for disabling buttons or clearing errors:

```js

formValidators[ profileForm.getAttribute('name') ].resetValidation()

// or you can use a string – the name of the form (you know it from `index.html`)

formValidators['profile-form'].resetValidation()

```

gennady-bars commented on src/Components/PopupWithForm.js:
```diff
@@ -22,9 +24,17 @@ export default class PopupWithForm extends Popup {
+            })
+            .finally(() => {
+                this.close();
+                this._popupSubmitButton.textContent = this._popupSubmitText;
```
Good job that you return the button text in `finally`

gennady-bars reviewed (commented):
Dear Brian, (**Please, expand the general comment ↓**)

Good job but there are some remarks:



* Please, make the avatar modal match the design on 320px (mobile) http://joxi.ru/J2bzbWltMQQoB2
* The code of `ok` checking is repeated in **every** request of the project. To avoid it you need to make a special method `_checkResponse` that will check any response from the server if it’s ok or not.
* I don't see some code of class `Api` . Actually you shouldn't have merged the changed of the 9th sprint into `main` unless the project has been accepted
* It’s better not to use `catch` in `Api` because you need to `catch` possible errors at the very **end** of any request and this end is in `index.js`   after all blocks `then`. And in real projects blocks `catch` usually open error popups and so on.
* You need to **toggle likes** only in block `then` after a successful response from the server because if there is an error the user will not know about that and will see the toggled heart color.
* You need to `disable` the button only after **resetting** the form when you have submitted a card in block `then`  to be able to click it again if there is a server error
* From the 8th sprint all constants (`initialCards` and `config`) should be in file `constants.js` in folder `utils` to clean up the main file.
* You need to `catch` possible errors at the very **end** of any server request.
* Class `UserInfo` should set the entire user’s data (including the **avatar**)


Please, correct everything and your project will be accepted.

Good luck with the refactoring

Brian-J-Wang committed:
refactor: API class loc decreased and improved readability

Brian-J-Wang committed:
refactor: clean up API class

Brian-J-Wang committed:
fix: avatar modal fixed in mobile view

Brian-J-Wang committed:
fix: cards are now rendered in the order in which they appear prior to refreshing the page
fix: catch errors are now completed in the index while promise rejects happens in the API class
refactor: constants are now moved to the constants file in the utils folder

Brian-J-Wang committed:
fix: popup with confirm class now shows 'deleting' instead of 'saving'
refactor: remove console.log in index.js

Brian-J-Wang committed:
fix: Popup with confirm now shows 'Deleting...' instead of 'closing'

gennady-bars commented on src/Components/Api.js:
```diff
@@ -1,62 +1,77 @@
+        });
     }
 
     handleError(res) {
```
You don't use `handleError` anymore. Please, delete it

gennady-bars commented on src/pages/index.js:
```diff
@@ -5,52 +5,57 @@ import UserInfo from "../Components/UserInfo.js";
+            API.addLike(data._id)
+            .then(() => {
+                data.isLiked = true;
+                heartElement.classList.add('places__heart_enabled');
```
`                heartElement.classList.add('places__heart_enabled');
`

Now only class `Card` knows every **card selector**, so you need to encapsulate the likes logics inside its methods and call them in `index.js`. It’s a violation of **Single responsibility** principle

gennady-bars commented on src/pages/index.js:
```diff
@@ -5,52 +5,57 @@ import UserInfo from "../Components/UserInfo.js";
+            API.removeLike(data._id)
+            .then(() => {
+                data.isLiked = false;
+                heartElement.classList.remove('places__heart_enabled');
```
**Could be improved**

To have access to any method of the class in `index.js` you can pass `this` to the call:

```js

        this._likeButton.addEventListener('click', () => {
            this._handleLike(this);
        });


```

And now in `index.js` you can call the methods like this:


```js


function addLike(cardInstance) {
        api.likeCard(cardInstance.getId())
            .then((res) => {
               cardInstance.addLike(res)
            })
            .catch((err) => {
                console.error(err);
            });
}


```

The same thing is with deleting cards

Brian-J-Wang replied:
Then I can use a similar function for removing likes and deleting the card as well.

gennady-bars reviewed (commented):
Everything is almost fine but:


* You need to **`disable`** the submit button in the `card popup` if the inputs are **empty** when you open it. It’s needed because I can click the button and add an empty card but it should **not** happen. After the first submit the button gets activated and you need to **deactivate** it again. Even if I can't add an empty card the button should look **disabled** (gray) anyway and with attribute `disabled`.
* Now only class `Card` knows every **card selector**, so you need to encapsulate the likes logics inside its methods and call them in `index.js`. It’s a violation of **Single responsibility** principle

Brian-J-Wang committed:
cleanup: removed handleError function, was not being used anymore
refactor: moved catch function to the card.js class

Brian-J-Wang committed:
fix: submit button is now properly disabled when reopened for the second time

gennady-bars reviewed (approved):
You’ve done a great job.

Your project has been accepted. 

Good luck
